src/lua/protobuf.lua

```diff
@@ -0,0 +1,642 @@
+        error(('Input cdata value %q for %q field is not integer'):format(
+            ffi.typeof(value), field_def.name))
+    end
+end
+
+-- }}} Global helpers
+
+
+-- {{{ is_scalar, is_enum, is_message
+
+local function is_scalar(field_def)
+    return scalars[field_def.type]
+end
+
+local function is_enum(protocol, field_def)
+    return protocol[field_def.type].type == 'enum'
+end
+
+local function is_message(protocol, field_def)
+    return protocol[field_def.type].type == 'message'
+end
+
+-- }}} is_scalar, is_enum, is_message
+
+
+-- {{{ Validation
+
+local function validate_length(value)
+    local MAX_LEN = 2^32
+    if string.len(value) > MAX_LEN then
+        error("Too long string to be encoded")
+    end
+end
+
+local function validate_table_is_array(field_def, value)
+    assert(type(value) == 'table')
+    local key_count = 0
+    local min_key = math.huge
+    local max_key = -math.huge
+    for k, _ in pairs(value) do
+        if type(k) ~= 'number' then
+            error(('Input array for %q repeated field ' ..
+                'contains non-numeric key: %q'):format(field_def.name, k))
+        end
+        if k - math.floor(k) ~= 0 then
+            error(('Input array for %q repeated field contains ' ..
+                'non-integer numeric key: %q'):format(field_def.name, k))
+        end
+        key_count = key_count + 1
+        min_key = math.min(min_key, k)
+        max_key = math.max(max_key, k)
+    end
+    if key_count == 0 then
+        return
+    end
+    if min_key ~= 1 then
+        error(('Input array for %q repeated field got min index %d. ' ..
+            'Must be 1'):format(field_def.name, min_key))
+    end
+    if max_key ~= key_count then
+        error(('Input array for %q repeated field has inconsistent keys. ' ..
+            'Got table with %d fields and max index of %d'):format(
+            field_def.name, key_count, max_key))
+    end
+end
+
+local function validate_type(field_def, value, exp_type)
+    if type(exp_type) == 'table' then
+        local found = false
+        for _, exp_t in pairs(exp_type) do
+            if type(value) == exp_t then
+                found = true
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value.'):format(
+            field_def.name, field_def.type, type(value)))
+    end
+    return
+end
+
+local function validate_range(field_def, value, range)
+    local min = range ~= nil and range[1] or nil
+    local max = range ~= nil and range[2] or nil
+    -- To avoid incorrect comparison unsigned cdata should not be
+    -- checked for lower limit. If one of the limits is 'nil' this
+    -- function skips the comparison with this limit.
```

Comment:
I left this sentence after @ImeevMA question why do we use `nil` to skip the comparison: https://github.com/tarantool/tarantool/pull/9488#discussion_r1637723173, as an example why actually we are skipping the comparison instead of comparison with `inf`. I will remove first sentence if it looks obcolete to you.

Reply:
Thanks for the clarification! It seems for me that the idea of the sentence is to explain why certain alternative implementation doesn't works. But it doesn't describe what is the implementation. For example, the following comment that aims the same goal should be clear for a reader without our context:

> If one of the limits is 'nil' this function skips the comparison with this limit.
>
> NB: We can't use `-math.huge` instead of `nil` for the lower limit, because, for example, `10ULL < -max.huge`.

It is OK for me to expand the sentence to make it more clear or to drop it.

Reply:
Added additional note after the comment